src/watchdog/observers/fsevents.py

```diff
@@ -67,86 +71,125 @@ class FSEventsEmitter(EventEmitter):
     def __init__(self, event_queue, watch, timeout=DEFAULT_EMITTER_TIMEOUT):
         EventEmitter.__init__(self, event_queue, watch, timeout)
         self._lock = threading.Lock()
+        # a dictionary of event.path -> posix.stat_result
+        self.filesystem_view = {}
 
     def on_thread_stop(self):
         if self.watch:
             _fsevents.remove_watch(self.watch)
             _fsevents.stop(self)
             self._watch = None
+        self.filesystem_view.clear()
+
+    def queue_event(self, event):
+        logger.info("queue_event %s", event)
+        EventEmitter.queue_event(self, event)
+
+    def queue_events(self, timeout, events):
+        i = 0
+        while i < len(events):
+            event = events[i]
+            logger.info(event)
+            src_path = self._encode_path(event.path)
+
+            """
+            FIXME: It is not enough to just de-duplicate the events based on
+                   whether they are coalesced or not. We must also take into
+                   account old and new state. As such we need to track all
+                   events that occurred so that we can make a correct decision
+                   about which events should be generated.
+            """
+
+            # For some reason the create and remove flags are sometimes also
+            # set for rename and modify type events, so let those take
+            # precedence.
+            if event.is_renamed:
+                # Internal moves appears to always be consecutive in the same
+                # buffer and have IDs differ by exactly one (while others
+                # don't) making it possible to pair up the two events coming
+                # from a singe move operation. (None of this is documented!)
```

Comment:
After some testing of the new emitter for my sync application, I have noticed that this detection of moves is not reliable: it seems to fail if one of the two moves is a coalesced event. I guess there is nothing we can do about it...

Reply:
Another case where this fails is a rename that changes only the casing of the name (APFS and HFS+ are not case-sensitive). This appears to be reported as two consecutive created events, the first with the old casing and the second with the new casing, which have event IDs that differ by 2.

Reply:
> Thank you @CCP-Aporia for the explanations.
> 
> > However, it's also to avoid an additional FileCreatedEvent in, for example, test_emitter.py:test_modify. There might be a simpler approach, but I haven't found it yet.
> 
> Hm, I have just reproduced this myself and I'm a bit puzzled by the behaviour. Why is this additional `FileCreatedEvent` emitted here? It doesn't event seem to be coalesced, it looks like a plain and simple and wrong created event.

As far as I have observed then the additional `FileCreatedEvent` happens because the `touch` operation after the `start_watching` call is almost always resulting in a coalesced event with the modify and create flags set.

Reply:
Thanks for testing this with your sync app! I'll take a look this week. It sounds like there are a couple of test cases missing that I need to add. 🙂

Reply:
Could we possible use the inode information from `filesystem_view` to better identify moves? If it's possible to do so without sacrificing performance, it might be an alternative to relying on relatively robust but undocumented behaviour. On the downside, it will further increase our dependence on manually querying the filesystem state.